server/src/daemon/prompter.rs

```diff
@@ -0,0 +1,210 @@
+        #[zbus(connection)] connection: &zbus::Connection,
+        #[zbus(header)] header: Header<'_>,
+    ) -> fdo::Result<()> {
+        println!("prompter: {}", header.path().unwrap());
+
+        let mut properties = properties.clone();
+
+        if header.path().unwrap().as_str().to_string().contains("/u") {
+            // setting properties related to Secret.Service.Unlock
+            properties.continue_label = Some(String::from("Unlock"));
+            properties.warning = Some(String::new());
+            properties.choice_chosen = Some(true);
+            properties.description = Some(String::from(
+                "An application wants access to the keyring \"login\", but it is locked.",
+            ));
+            properties.title = Some(String::from("Unlock Keyring"));
+            properties.message = Some(String::from("Authentication required"));
+            properties.choice_label = Some(String::from(
+                "Automatically unlock this keyring whenever I'm logged in",
+            ));
+            properties.caller_window = Some(String::new());
+            properties.cancel_label = Some(String::from("Cancel"));
+        } else {
+            // setting properties related to Secret.Service.CreateCollection
+            properties.continue_label = Some(String::from("Continue"));
+            properties.warning = Some(String::new());
+            properties.choice_chosen = Some(false);
+            properties.description = Some(String::from(
+                "An application wants to create a new keyring. Choose the password you want to use for it.",
+            ));
+            properties.title = Some(String::new());
+            properties.message = Some(String::from("Choose password for new keyring"));
+            properties.choice_label = Some(String::new());
+            properties.caller_window = Some(String::new());
+            properties.cancel_label = Some(String::from("Cancel"));
+            properties.password_new = Some(true);
+        }
+
+        *self.properties.lock().unwrap() = properties.clone();
+
+        let path = Arc::new(header.path().unwrap().to_owned());
+        let exchange = exchange.to_string();
+        let connection = Arc::new(connection.to_owned());
+
+        tokio::spawn(async move {
+            let prompter = PrompterProxy::new(&connection).await.unwrap();
+            prompter
+                .perform_prompt(&path, "password", properties, exchange)
+                .await
+                .unwrap();
+        });
+
+        Ok(())
+    }
+
+    pub async fn prompt_done(
+        &self,
+        #[zbus(object_server)] object_server: &zbus::ObjectServer,
+    ) -> fdo::Result<()> {
+        object_server.remove::<Self, _>(&self.path).await?;
+        Ok(())
+    }
+}
+
+impl PrompterCallback {
+    pub fn new(post_fix: Option<&str>, prompts_counter: i32) -> Self {
+        let path = if post_fix.is_some() {
```

Comment:
Same thing, two constructors or take the `PromptSource` argument.

Reply:
Done